misk/src/main/kotlin/misk/web/RealNetworkChain.kt

```diff
@@ -0,0 +1,35 @@
+package misk.web
+
+import misk.NetworkChain
+import misk.NetworkInterceptor
+import misk.web.actions.WebAction
+import kotlin.reflect.KFunction
+
+internal class RealNetworkChain(
+  private val _action: WebAction,
+  private val _request: Request,
+  private val interceptors: List<NetworkInterceptor>,
+  private val _function: KFunction<*>,
+  private val index: Int = 0
+) : NetworkChain {
+
+  override val action: WebAction
+    get() = _action
+
+  override val request: Request
+    get() = _request
+
+  override val function: KFunction<*>
+    get() = _function
+
+  override fun proceed(request: Request): Response<*>? {
+    check(index < interceptors.size) { "final interceptor must be terminal" }
+    // NB(young): At this point it's possible the Request has been manipulated by Network
```

Comment:
I don't think we want to allow network interceptors to mutate ``Request`` properties

Reply:
I’ve found it to be occasionally profitable. In particular, a network interceptor could gunzip an incoming request body.

Reply:
though I do think we should not re-route; that’s not good. If we were superparanoid we could check to see if the route is still valid and crash with an IllegalStateException if it doesn’t

Reply:
I was thinking it could potentially be useful if you're behind a proxy and need to move headers around, eg `X-Real-IP` or something.  I don't think we'll need to use it anytime soon anyway.